pkg/reconciler/buildlimitcleanup/controller.go

```diff
@@ -0,0 +1,134 @@
+	ctx = ctxlog.NewContext(ctx, "build-limit-cleanup-controller")
+	return add(ctx, mgr, NewReconciler(c, mgr), c.Controllers.Build.MaxConcurrentReconciles)
+}
+func add(ctx context.Context, mgr manager.Manager, r reconcile.Reconciler, maxConcurrentReconciles int) error {
+	// Create the controller options
+	options := controller.Options{
+		Reconciler: r,
+	}
+
+	if maxConcurrentReconciles > 0 {
+		options.MaxConcurrentReconciles = maxConcurrentReconciles
+	}
+
+	// Create a new controller
+	c, err := controller.New("build-limit-cleanup-controller", mgr, options)
+	if err != nil {
+		return err
+	}
+
+	pred := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			o := e.Object.(*buildv1alpha1.Build)
+			return o.Spec.Retention != nil && (o.Spec.Retention.FailedLimit != nil || o.Spec.Retention.SucceededLimit != nil)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			n := e.ObjectNew.(*buildv1alpha1.Build)
+			o := e.ObjectOld.(*buildv1alpha1.Build)
+
+			/* Check to see if there are new retention parameters or whether the
+			limit values have decreased */
+			if o.Spec.Retention == nil && n.Spec.Retention != nil {
+				if n.Spec.Retention.FailedLimit != nil || n.Spec.Retention.SucceededLimit != nil {
+					return true
+				}
+			} else if n.Spec.Retention != nil && o.Spec.Retention != nil {
+				if n.Spec.Retention.FailedLimit != nil && o.Spec.Retention.FailedLimit == nil {
+					return true
+				} else if n.Spec.Retention.SucceededLimit != nil && o.Spec.Retention.SucceededLimit == nil {
+					return true
+				} else if n.Spec.Retention.FailedLimit != nil && o.Spec.Retention.FailedLimit != nil && int(*n.Spec.Retention.FailedLimit) < int(*o.Spec.Retention.FailedLimit) {
+					return true
+				} else if n.Spec.Retention.SucceededLimit != nil && o.Spec.Retention.SucceededLimit != nil && int(*n.Spec.Retention.SucceededLimit) < int(*o.Spec.Retention.SucceededLimit) {
+					return true
+				}
+			}
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// Never reconcile on deletion, there is nothing we have to do
+			return false
+		},
+	}
+
+	predBuildRun := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			// Never reconcile in case of create buildrun event
+			return false
+		},
+		// Reconcile the build the related buildrun has just completed
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			n := e.ObjectNew.(*buildv1alpha1.BuildRun)
+
+			// check if Buildrun is related to a build
+			if n.Spec.BuildRef == nil {
+				return false
+			}
+
+			o := e.ObjectOld.(*buildv1alpha1.BuildRun)
+			oldCondition := o.Status.GetCondition(buildv1alpha1.Succeeded)
+			newCondition := n.Status.GetCondition(buildv1alpha1.Succeeded)
+			if oldCondition != nil && newCondition != nil {
+				if (oldCondition.Status == corev1.ConditionUnknown) &&
+					(newCondition.Status == corev1.ConditionFalse || newCondition.Status == corev1.ConditionTrue) {
+					return true
+				}
+			}
```

Comment:
We here need to handle the case that the old version has no succeeded condition and the new one is already completed (could happen if we fail in the validation phase and never really start the BuildRun).

```suggestion
			if newCondition != nil {
				if (oldCondition == nil || oldCondition.Status == corev1.ConditionUnknown) &&
					(newCondition.Status == corev1.ConditionFalse || newCondition.Status == corev1.ConditionTrue) {
					return true
				}
			}
```